### DISCUSSION TITLE
About Travel Time (duration) calculation

### DISCUSSION THREAD

**Discussion by mertbakir** [valhalla/valhalla#3354] [Q&A] [answered]
Hi, probably, the distance multiplied by some coefficient based on the avg speed for that way in the OSM and cost profile to find the travel time.  I wonder if we can pass avg speed for a vehicle (costing profile) via API request.
In the docs, I'm looking at [speeds](https://github.com/valhalla/valhalla/blob/master/docs/speeds.md), [costing options](https://valhalla.readthedocs.io/en/latest/api/turn-by-turn/api-reference/). Yet, couldn't get if it's possible to pass the parameter in the API request or these values are fixed values in the OSM file?

**Comment by kevinkreiser** [comment, marked as answer]
Currently we support many different ways of storing speed information about an edge in the graph. For most of the motorized vehicles we use this speed to determine the amount of time it takes to traverse a given edge. The different types of speed are: real time from traffic, historical from traffic, speed limit, statistical or heuristic when those other sources dont exist. The 3rd category I listed are speeds which come either directly or indirectly from osm. Some costing models do allow you to set a max speed so that when they are on motorways we dont overestimate thier speed but we dont allow setting an average speed because for most motorized vehicles the speed will be variable. If really want to use a single speed average you can just use the distances in the response and do it yourself on the client side. That said we do allow setting walking and biking speeds, but we modulate that speed based on the properties of a given edge in the graph, steep uphilll or unpaved roads will be slower to traverse for example.

**Reply by Bondify** [reply]
hi @kevinkreiser , first of all, thanks a lot for the support you give to Valhalla, it is really world class! 

I am hosting my own Valhalla server and would like to use the route API to get runtimes to finally get the car speeds by street edge. I've been reading through the API's doc but I don't quite understand what the source of the speeds is. You mention different types of speeds (real time from traffic, historical from traffic, speed limit, statistical, etc).

Is this true even if I'm hosting my own Valhalla server? Or when hosting my own server it would always take one type of speed (let's say speed limit)? 

Thanks a lot for your help,
Santiago

**Reply by Bondify** [reply]
I'm mostly asking because in the docs I found that the speed is only assigned depending on the OSM tag: https://valhalla.github.io/valhalla/speeds/

**Reply by nilsnolde** [reply]
No traffic influence of course for vanilla OSM deployments.

Statistical: https://github.com/OpenStreetMapSpeeds/. There's multiple references in this repo how to use that.
Heuristic: correct, mainly OSM derived from the highway tag (and some other props like local density, some penalties etc)

**Reply by Bondify** [reply]
thanks for the quick reply @nilsnolde ! What would be the opposite of a "OSM vanilla deployments"? 

Sorry if it is a silly question, I'm not very familiar with this :)

**Reply by nilsnolde** [reply]
E.g. integrating some commercial live traffic service (TomTom for example). Or go all in and buy a TomTom network, convert that to OSM format for Valhalla to even accept it, so you can also use their historical traffic data (+/- live traffic).

**Reply by mertbakir** [reply]
> E.g. integrating some commercial live traffic service (TomTom for example). Or go all in and buy a TomTom network, convert that to OSM format for Valhalla to even accept it, so you can also use their historical traffic data (+/- live traffic).

May I ask why would anyone buy TomTom network instead of using their APIs? What would be the use case?

**Reply by nilsnolde** [reply]
Costs mostly when you start scaling for Mios of requests, apart from using world class open software that you can customize to your needs😄 for many commercial applications it’s vital to be in control of your tech stack.

**Reply by Bondify** [reply]
thanks @nilsnolde ! I think we have some TomTom network data. I might look into how to format it to OSM format, although it seems like a very complex thing from what I've been reading in the forum.

**Reply by nilsnolde** [reply]
I should mention that we offer that as a service/software to license: https://github.com/gis-ops/prop2osm.

**Reply by Bondify** [reply]
that's very cool @nilsnolde , thanks a lot! I'll check with my team if we want to give this a shot :)
